### PR TITLE
Prevent LegacyGlobalOptionService from GC rooting PreviewWorkspace

### PR DESCRIPTION
**Pull request by sharwell**
Fixes [AB#1756080](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1756080)

### CODE REVIEW

**Review by tmat — APPROVED** [review]

**Review by jasonmalinowski — CHANGES_REQUESTED** [review]
This seems to be because we're not disposing the PreviewWorkspace which would have done the unsubscribe. Although this may be a good defense in depth, can we fix that too?

**Comment by jasonmalinowski** [comment]
Or alternatively: if we're making this weak, can we just get rid of the Unregister method entirely?

**Comment by sharwell** [comment]
> This seems to be because we're not disposing the PreviewWorkspace which would have done the unsubscribe. Although this may be a good defense in depth, can we fix that too?

Ideally yes. In practice, the cancellation paths involved in ensuring that happens are _extra_ non-trivial.

**Comment by jasonmalinowski** [comment]
@sharwell: my concern isn't the unregister list specifically, but there's a lot of _other_ stuff also happening in Workspace.Dispose() that we're also skipping, and I have no idea if any of that is safe to skip or whether that's also causing additional rooting. Basically, we're violating the contract of the Workspace, we got caught, and this isn't fixing that.

If we're going to take this approach due to us servicing this for 17.5, then I'd request a few additional things:

1. A unit test verifying this actually works and there isn't something else rooting it. Given we already have helpers for writing tests like this I don't believe this should be difficult to write. The chance of this being regressed by accident is _very_ high.
2. A comment in the Dispose() method explaining that people really shouldn't put any new event unsubscriptions in there.
3. A tracking bug to either completely empty the Dispose() method or fix the fact PreviewWorkspace is not being disposed -- either is fine with me.

I actually don't really like the idea that Workspace is Disposable so I don't mind the direction this PR takes, but this PR as written leaves us in a place that's likely to have regressions without a test, and we should fix that.

File: src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs [hunk 1]
```diff
@@ -238,6 +238,19 @@ public void TestPreviewWorkspaceDoesNotLeakSolution()
             solutionObjectReference.AssertReleased();
         }
 
+        [Fact]
+        public void TestPreviewWorkspaceDoesNotLeakItself()
+        {
+            var composition = EditorTestCompositions.EditorFeatures;
+            var exportProvider = composition.ExportProviderFactory.CreateExportProvider();
+            var previewWorkspaceReference = ObjectReference.CreateFromFactory<Workspace, TestComposition>(
+                static composition => new PreviewWorkspace(composition.GetHostServices()),
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs:247]: Was there a reason the static lambda was necessary here? We don't care about allocations, was this needed for correctness?

File: src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs [hunk 2]
```diff
@@ -238,6 +238,19 @@ public void TestPreviewWorkspaceDoesNotLeakSolution()
             solutionObjectReference.AssertReleased();
         }
 
+        [Fact]
+        public void TestPreviewWorkspaceDoesNotLeakItself()
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs:242]: Oh nice adding tests for workspaces generally and also PreviewWorkspace specifically.

File: src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs [hunk 3]
```diff
@@ -238,6 +238,19 @@ public void TestPreviewWorkspaceDoesNotLeakSolution()
             solutionObjectReference.AssertReleased();
         }
 
+        [Fact]
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs:241]: Consider adding the work item attribute to link to the bug here so if anybody wonders why we're asserting this it's clear it was for memory leak reasons. (I can imagine somebody going "oh, this is a stupid test, people should call Dispose()" and we're back here again)

**Review by jasonmalinowski — APPROVED** [review]

File: src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs:252]: I wanted to make sure there was no implicit state machine or capture created for this method, so I know where all locals live.

File: src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs (hunk 3 above)
Comment (sharwell) [reply src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs:246]: Will amend the last comment to add comments and work item links.

**Comment by sharwell** [comment]
> A comment in the Dispose() method explaining that people really shouldn't put any new event unsubscriptions in there.

I'm not sure this matters in practice. If someone is thinking about adding an unsubscribe to that method, it seems unlikely that a comment would have any impact on the outcome.

**Comment by jasonmalinowski** [comment]
> If someone is thinking about adding an unsubscribe to that method, it seems unlikely that a comment would have any impact on the outcome.

Even if the tests would catch them I can still imagine people asking "so are workspaces supposed to be disposed or not?" and there's not a clear answer either way.

**Comment by sharwell** [comment]
I'm going to get this merged since it fixes the one known issue. Hopefully in the future we can fix other cases we find where solutions are not disposed, but at least we won't be leaking memory from them.

**Pull request merged**